### ISSUE TITLE
deblender does not deblend

### ISSUE THREAD

**Issue by esheldon** [esheldon/metadetect#123]
I suspect we need to set `deblend_config.propagateAllPeaks = True` but need to wait for confirmation from Bob

**Comment by esheldon** [comment]
I think this also means we need two separate options

* deblend (to run the deblender at all)
* use_noise_replacer (run the noise replacer)

I'm not sure what the noise replacer actually does but I expect it replaces neighboring footprints with noise rather than neighboring objects in a blend.  I'm not sure we want to do that at all.

**Comment by esheldon** [comment]
This represents a bit if a quandry because if we don't run the deblender we only get a single peak returned.  In that case we are missing some objects, which we would never do.

I'm not sure how to test the case of "we found these peaks but aren't going to deblend"

**Comment by esheldon** [comment]
I think I've pieced together an explanation

- If you don't run deblending it will find the sub-peaks, but the source list will only contain one entry.  There might be a way to recover the sub-peaks but I don't know how.  If we can do that we can test with and without deblending but using all objects.
- If you run deblending it will have a parent plus N children (we ignore the parent)
- If you run a noise replacer it will set neighbors to noise and then use the symmetry of the object of interest to add in signal in that region

**Comment by esheldon** [comment]
@rearmstr does this make sense?

**Comment by esheldon** [comment]
Correction:  if you don't set `deblend_config.propagateAllPeaks = True` deblending sometimes occurs, but it seems to be ineffecitve.

So I think the thing to do is always run the deblend task, with that setting, and then the option will be to "deblend" as we think of it, which means run the noise replacer.

For clarity I propose that we make the option in the codes (and config) simply "deblend"

But I want to get confirmation from @rearmstr before making any moves

**Comment by esheldon** [comment]
I've got a branch `deblend` going with an implementation of the above

**Comment by rearmstr** [comment]
I don't know if this is the right solution. Certainly in the short term it fixes the problem. But I would like to understand why the deblender is skipping this object. It shouldn't need to be forced to deblend it.  I haven't seen this behavior before running on very similar types of images, but I have been using older versions of the stack.

I haven't been able to figure out the root of the problem yet, so maybe I should ask the scarlet folks.

**Comment by esheldon** [comment]
OK, thanks for looking Bob

**Comment by rearmstr** [comment]
I think I found the problem. When computing the psf image (psf.computeImage) the coordinate system is centered on the psf and not in the image coordinates. There is a place in the deblender where it looks for the overlap of the psf image and the footprint bounding box.  If there is no overlap then it will skip the object. It doesn't appear to keep this information anywhere so it was hard to find.

So it looks like the custom PSF defined in descwl_sims needs to be modified.

**Comment by beckermr** [comment]
I don't follow. Does this mean the stack can't ever run with a constant psf?

**Comment by rearmstr** [comment]
No, this code is not using a official dm PSF. It is trying to build a class to have all the same functions as a dm PSF. I'm saying that the image produced when running this custom computeImage function (https://github.com/LSSTDESC/descwl-shear-sims/blob/master/descwl_shear_sims/psfs/dmpsfs.py#L60) is not in the right coordinate system.

**Comment by beckermr** [comment]
Ah.  Thank you. That makes sense.

**Comment by esheldon** [comment]
Do you mean this code?
```python
        origin = -(dim//2)
        dims = (dim, )*2
        bbox = geom.Box2I(geom.Point2I(origin), geom.Extent2I(dims))
```

**Comment by rearmstr** [comment]
While I understand how much it simplifies things, i think having a custom PSF class is a bit risky. We have no way of tracking if the internals of the dm psf code change. They have been mostly stable for the past few years, but that doesn't mean they won't change in the future. 

There is a straight forward way to create a constant dm-compatible psf, but anything more complicated is more difficult.

**Comment by esheldon** [comment]
I believe I stole that from a class Josh Meyers had written to wrap a galsim object to make it produce a dm psf

**Comment by esheldon** [comment]
Agreed this is risky, but we didn't know what else to do, particularly for the non-constant power spectrum psf

**Comment by esheldon** [comment]
Could you help us put together the "official" one that works for a fixed psf image?  I know this code but presumably it will also not work

```python
    return KernelPsf(
        FixedKernel(
            afw_image.ImageD(psf_image.astype(float))
        )
    )
```

**Comment by esheldon** [comment]
Actually using the above code does work

**Comment by rearmstr** [comment]
Oh, actually I realized that I have the same issue in some of my sims where I use a KernelPSF. So I'm not sure if it will work to compute a fixed psf image. I hadn't seen it before because I am not running the deblender stage where this is triggered.

I probably won't have much time before the end of september due to the end of the fiscal year, but I could take a look in October.

**Comment by esheldon** [comment]
Well, for this particular setup it does seem to work, but I couldn't guess why it works here and not for your sims.

**Comment by rearmstr** [comment]
Wait, I'm wrong. It does work for my sims. I forgot that if you don't give computeImage an argument it will use a default point that is class dependent as the point to compute the image at. It was defaulting to the point 0,0, so I thought the coordinate system was centered on the PSF, but if I give it a different coordinate it is in the right coordinate system. (I think Josh has an RFC to change this behavior in the stack)

**Comment by esheldon** [comment]
Note for the cell based coadds we produce, we store a KernelPsf like above.  This is because we have created a coadded psf, where the psf images we coadd are drawn at locations corresponding to the coadd center.  So the coadd psf is not stored as a spatially varying psf.

I ran some tests and indeed the deblending seems to be working if I run all they way through to the coadd.

**Comment by rearmstr** [comment]
Do we expect to be able to treat the PSF as constant for the cell sizes you have been using? Does this depend on the number of images in the coadd?

**Comment by beckermr** [comment]
yes we will treat it as constant and we will always have enough epochs

**Comment by esheldon** [comment]
Our passed each other in flight

Yes, I think treating the psf as constant is fine over these sizes, assuming our power spectrum psf has realistic variations.

In the metadetect code where I run detection, it is going to be seeing the KernelPsf.  Also when we use the psf for metacal image deconvolution I call computeKerneImage.

Do you expect all that to do the right thing?

**Comment by rearmstr** [comment]
Yes, I think that should be fine.

**Comment by esheldon** [comment]
OK.  Then our fake psfs are probably not a problem, the way we are using them now.

But I will go ahead and convert our code to just create a fixed kernel psf for the constant case.

We should think about how we have faked the varying dm psf though.

**Comment by rearmstr** [comment]
Yeah, the fake psfs should work once the coordinate system is fixed. I guess there are some things they can't do like be persisted with the exposure, but those probably don't matter for the way we are using them. 

My only worry is a couple years from now suddenly the simulation codes starts failing or gets different results because of some internal dm change. (This has already happened to me on a couple of occasions). But probably not a huge issue.

**Comment by esheldon** [comment]
Well, we can't actually use KernelPsf for the constant psfs because it doesn't do sub-pixel offsets.

But in any case I think we are ok downstream, but it messes up simple tests like this.  So we should probably bite the bullet and fix the bounding box.

**Comment by rearmstr** [comment]
I think the KernelPsf should allow for sub-pixel offset.s

**Comment by esheldon** [comment]
ok, I thought I tested this

**Comment by rearmstr** [comment]
How were you testing it?

**Comment by esheldon** [comment]
I thought I tested using `computeImage`, which supports the shifting of the image, and either it wasn't supported or it just gave an unshifted image

**Comment by rearmstr** [comment]
I just tried to double check this by doing some simple tests. It looks like computeImage does shift the image values as expected depending on the location passed to the argument.

**Comment by esheldon** [comment]
I think I remember now what the issue was.

To get the right psf for the coadd, we need to start with shifted PSFs and then shift them back when we coadd them.

That initial shifting needs to be without any extra smearing.  I think if I implemented a fixed psf and had it shift it over, and then back for coadding, it would get smeared twice.
